client/main.go

```diff
@@ -17,93 +17,105 @@ func main() {
 	if err != nil {
 		panic(err)
 	}
-	reader := bufio.NewReader(conn)
-	serverHandlers := make(map[scs.ActionName]ServerHandleFunc)
-	serverHandlers[scs.GetRooms] = handleGetRooms
-	serverHandlers[scs.Post] = handleMessage
-	serverHandlers[scs.JoinRoom] = handleJoinRoom
-	serverHandlers[scs.LeaveRoom] = handleLeaveRoom
-	state := &State{state: INIT}
+	serverHandlers := map[chat.ActionName]ServerHandleFunc{
+		chat.GetRooms:   handleGetRooms,
+		chat.Post:       handleMessage,
+		chat.JoinRoom:   handleJoinRoom,
+		chat.LeaveRoom:  handleLeaveRoom,
+		chat.CreateRoom: handleCreateRoom,
+	}
+
+	state := &State{state: initial}
 
-	go handleServerInput(reader, serverHandlers, state)
-	rw := bufio.NewReadWriter(bufio.NewReader(os.Stdin), bufio.NewWriter(conn))
+	go handleServerInput(bufio.NewReadWriter(bufio.NewReader(conn), bufio.NewWriter(conn)), serverHandlers, state)
 
 	userHandlers := map[ClientState]map[string]UserHandleFunc{
-		INIT: {
+		initial: {
 			"create": createRoom,
 			"join":   joinRoom,
 		},
-		IN_ROOM: {
+		inRoom: {
 			"post":  post(state),
 			"leave": leaveRoom,
 		},
 	}
-
-	loop(rw, state, userHandlers)
+	acceptUserInput(bufio.NewReadWriter(bufio.NewReader(os.Stdin), bufio.NewWriter(conn)), state, userHandlers)
 }
 
 type State struct {
 	state ClientState
-	room  scs.Room
+	room  chat.Room
 }
 
 type ClientState int
 
 const (
-	INIT    ClientState = 0
-	IN_ROOM ClientState = 1
+	initial ClientState = 0
+	inRoom  ClientState = 1
 )
 
-type ServerHandleFunc func(*bufio.Reader, *State)
+type ServerHandleFunc func(*bufio.Reader, func(ClientState, chat.Room))
 type UserHandleFunc func(*bufio.Writer, string)
 
-func handleServerInput(reader *bufio.Reader, handlers map[scs.ActionName]ServerHandleFunc, state *State) {
+func handleServerInput(rw *bufio.ReadWriter, handlers map[chat.ActionName]ServerHandleFunc, state *State) {
+	updateStateFunc := updateState(state, rw)
 	for {
-		action := &scs.Action{}
-		if err := gob.NewDecoder(reader).Decode(action); err != nil {
+		action := &chat.Action{}
+		if err := gob.NewDecoder(rw).Decode(action); err != nil {
 			if err == io.EOF {
 				log.Println("Reached EOF - close this connection.\n   ---")
-			} else {
-				log.Println("\nError reading command.", err)
+				return
 			}
+			chat.LogError("Error reading command: %w", err)
```

Comment:
Would inline the code here. Also, `%w` is only meaningful for wrapping the error with `fmt.Errorf`. If you're only printing it, you can just do `log.Printf("foo: %v", err)`, or whatever.